### ISSUE TITLE
Android: How to overlap/overlay publisher view over subscriber view?

### ISSUE THREAD

**Issue by sunweiyang** [opentok/opentok-react-native#122]
We need to use `opentok-react-native` to create overlapping views on the same screen (publisher thumbnail over subscriber). 

Here is an example screen of what we need:

![image](https://user-images.githubusercontent.com/12490625/42005408-02a5ee4a-7a29-11e8-84cf-84a530740ffd.png)

We've tried to use `position: absolute` styling, but this fails to render properly:
```
import { Dimensions, View } from "react-native";
import { OTSession, OTPublisher, OTSubscriber } from "opentok-react-native";

const windowWidth = Dimensions.get("window").width;
const windowHeight = Dimensions.get("window").height;

...

          <View style={{ flex: 1 }} >
            <OTSession
              apiKey={this.apiKey}
              sessionId={this.sessionId}
              token={this.token}
              eventHandlers={this.sessionEventHandlers}
            >
              <OTSubscriber
                properties={this.subscriberProperties}
                eventHandlers={this.subscriberEventHandlers}
                style={{ width: windowWidth, height: windowHeight }}
              />
              <OTPublisher
                style={{
                  position: "absolute",
                  top: windowWidth / 10,
                  right: windowWidth / 20,
                  width: windowWidth / 5,
                  height: windowHeight / 6
                }}
                properties={this.publisherProperties}
                eventHandlers={this.publisherEventHandlers}
              />
            </OTSession>
          </View>
...
```

The straightforward non-overlapping implementation (like in `opentok-react-native-samples`) renders as expected.
How can our desired overlap/overlay of `OTPublisher` on top of `OTSubscriber` be accomplished?

**Comment by sunweiyang** [comment]
@jonathanwmaddison thanks, I can also confirm that absolute positioning of publisher works on iOS.

I've figured out that on Android, for a normal 2-user OpenTok session, the subscriber currently always be rendered on top of the publisher. I've observed that the publisher view instantiates and definitely exists for Android, but is simply covered up by the subscriber view. Unfortunately, defining `zIndex` style does not affect this behavior.

Does anyone have a working Android solution for this?

**Comment by jonathanwmaddison** [comment]
@sunweiyang Have you found any way around this issue? I haven't dug into the native android implementation since Android is currently lower on our priority list.

**Comment by sunweiyang** [comment]
@jonathanwmaddison Since we also have limited development resources, for now we have simply fallen back on pure Android native (before we began using opentok-react-native). This is currently the main blocker for us from using opentok-react-native in Android. If/when this issue is resolved, we'll immediately update our Android codebase for opentok-react-native.

**Comment by rockyip** [comment]
Hi, I have the same issue as yours and I found a solution from the android SDK
[https://github.com/opentok/opentok-android-sdk-samples/issues/136#issuecomment-322218524](url)
Just apply the change to the method "initSession" in OTSessionManager.java

**Comment by msach22** [comment]
@sunweiyang @jonathanwmaddison @rockyip Thanks for sharing your findings. I will check this out and get back to you all soon!

**Comment by asatour** [comment]
Hello, 
I managed to resolve this problem by using Modal (that I make visible only when the subscriber is connected)

```
const tokboxStyle =
      {
        publisher: {
          width: 100,
          height: 100,
          position: 'absolute',
          top: 220,
          left: 10,
          backgroundColor: 'red',
          zIndex: 30,
        },
        subscriber: {
          backgroundColor: 'yellow',
          width: 400,
          height: 400,
        },
      }

    const {tokboxProToken, tokboxSessionId, tokboxApiKey} = this.props.teleconsultation
```
```
    return (

      <View>
        <OTSession
          apiKey={tokboxApiKey}
          sessionId={tokboxSessionId}
          token={tokboxProToken}
          onError={this.onSessionError}
          eventHandlers={this.sessionEventHandlers}
        >
          {
            true ?
              (
                <View>
                  <Modal
                    animationType="slide"
                    transparent
                    visible={this.state.connection === this.STATUS.SUBSCRIBER_CONNECTED}
                  >
                    <OTPublisher
                      properties={{width: 100, height: 100}}
                      style={tokboxStyle.publisher}
                      onPublish={this.onPublish}
                      onError={this.onPublishError}
                      eventHandlers={this.publisherEventHandlers}
                    />
                  </Modal>
                  <OTSubscriber
                    properties={{width: 400, height: 400}}
                    style={tokboxStyle.subscriber}
                    onSubscribe={this.onSubscribe}
                    onError={this.onSubscribeError}
                    eventHandlers={this.subscriberEventHandlers}
                  />

                </View>
              ) :
              (<Text>Connecting</Text>)
          }

        </OTSession>
      </View>
     )
```

**Comment by asatour** [comment]
Hello, 
I managed to resolve this problem by using Modal (that I make visible only when the subscriber is connected)

```
renderTocKBox() {
    const tokboxStyle =
      {
        publisher: {
          width: 100,
          height: 100,
          position: 'absolute',
          top: 220,
          left: 10,
          backgroundColor: 'red',
          zIndex: 30,
        },
        subscriber: {
          backgroundColor: 'yellow',
          width: 400,
          height: 400,
        },
      }

    const {tokboxProToken, tokboxSessionId, tokboxApiKey} = this.props.teleconsultation
    return (

      <View>
        <OTSession
          apiKey={tokboxApiKey}
          sessionId={tokboxSessionId}
          token={tokboxProToken}
          onError={this.onSessionError}
          eventHandlers={this.sessionEventHandlers}
        >
          <View>
            <Modal
              animationType="slide"
              transparent
              visible={this.state.connection === this.STATUS.SUBSCRIBER_CONNECTED}
            >
              <OTPublisher
                properties={{width: 100, height: 100}}
                style={tokboxStyle.publisher}
                onPublish={this.onPublish}
                onError={this.onPublishError}
                eventHandlers={this.publisherEventHandlers}
              />
            </Modal>
            <OTSubscriber
              properties={{width: 400, height: 400}}
              style={tokboxStyle.subscriber}
              onSubscribe={this.onSubscribe}
              onError={this.onSubscribeError}
              eventHandlers={this.subscriberEventHandlers}
            />

          </View>
        </OTSession>
      </View>

    )
  }

```

**Comment by asatour** [comment]
You should draw the publisher again when the subscriber is viewed (so in two case: when the is connecting and when the subscriber is streaming
 ```
renderTocKBox() {
    console.info(this.state.connection)
    const {tokboxProToken, tokboxSessionId, tokboxApiKey} = this.props.teleconsultation
    return (

      <View>
        <OTSession
          apiKey={tokboxApiKey}
          sessionId={tokboxSessionId}
          token={tokboxProToken}
          onError={this.onSessionError}
          eventHandlers={this.sessionEventHandlers}
        >
          <View>
            {
              this.state.connection !== this.STATUS.CONNECTING || this.state.connection === this.STATUS.SUBSCRIBER_CONNECTED ?
                (
                  <View>
                    <OTPublisher
                      properties={styles.publisher.properties}
                      style={styles.publisher.style}
                      onPublish={this.onPublish}
                      onError={this.onPublishError}
                      eventHandlers={this.publisherEventHandlers}
                    />
                  </View>
                ) : null
            }

            <OTSubscriber
              properties={styles.subscriber.properties}
              style={styles.subscriber.style}
              onSubscribe={this.onSubscribe}
              onError={this.onSubscribeError}
              eventHandlers={this.subscriberEventHandlers}
            />
            {
              this.state.connection !== this.STATUS.SUBSCRIBER_CONNECTED ?
                (
                  this.renderConnecting()
                ) : null
            }
          </View>
        </OTSession>
      </View>

    )
  }
```

**Comment by sunweiyang** [comment]
Thanks @asatour but my app needs a modal already for other purposes, and RN currently does not support multiple simultaneous modals:
https://github.com/facebook/react-native/issues/3445
https://react-native.canny.io/feature-requests/p/multiple-modal-components

A true solution for overlapping publisher still needs to be found that does not use modals.

**Comment by asatour** [comment]
@sunweiyang 
I'm sorry but I removed the modal solution, If you like at the response now, it's without modal
I just draw the OTPublisher in the first time and when I get stream from the subscriber, look at the line code:

`this.state.connection !== this.STATUS.CONNECTING || this.state.connection === this.STATUS.SUBSCRIBER_CONNECTED ?`

**Comment by msach22** [comment]
@sunweiyang Sorry for the delay on this, I'll add a PR to fix this and release hopefully by today!

**Comment by msach22** [comment]
@sunweiyang @asatour @jonathanwmaddison @emin93 @sandrotiiago

This has been fixed and released with `v0.8.4`. Please let me know if you have any questions!

**Comment by msach22** [comment]
BTW, this was my code that I used to test:
```
  render() {
    const { height, width } = Dimensions.get('window');
    return (
      <View style={{ flex: 1}}>
        <OTSession apiKey={this.apiKey} sessionId={this.sessionId} token={this.token}>
          <OTPublisher style={{ 
            position: 'absolute', bottom: 20, right: 20, width: 100, height: 100, zIndex: 1000 }} />
          <OTSubscriber style={{ width, height }} />
        </OTSession>
      </View>
    );
  }
```
